crates/ruma-client-api/src/uiaa/get_uiaa_fallback_page.rs

```diff
@@ -35,35 +34,179 @@ pub mod v3 {
 
+    /// Response type for the `authorize_fallback` endpoint.
+    #[derive(Debug, Clone)]
+    #[non_exhaustive]
+    pub enum Response {
+        /// The response is a redirect.
+        Redirect(Redirect),
+
+        /// The response is an HTML page.
+        Html(HtmlPage),
+    }
+
+    /// The data of a redirect.
+    #[derive(Debug, Clone)]
+    #[non_exhaustive]
+    pub struct Redirect {
+        /// The URL to redirect the user to.
+        pub url: String,
+    }
+
+    /// The data of a HTML page.
+    #[derive(Debug, Clone)]
+    #[non_exhaustive]
+    pub struct HtmlPage {
+        /// The body of the HTML page.
+        pub body: Vec<u8>,
+    }
+
     impl Response {
-        /// Creates a new `Response` with the given HTML body.
-        pub fn new(body: Vec<u8>) -> Self {
-            Self { redirect_url: None, body }
+        /// Creates a new HTML `Response` with the given HTML body.
+        pub fn html(body: Vec<u8>) -> Self {
+            Self::Html(HtmlPage { body })
         }
 
-        /// Creates a new `Response` with the given redirect URL and an empty body.
+        /// Creates a new HTML `Response` with the given redirect URL.
         pub fn redirect(url: String) -> Self {
-            Self { redirect_url: Some(url), body: Vec::new() }
+            Self::Redirect(Redirect { url })
+        }
+    }
+
+    #[cfg(feature = "server")]
+    impl ruma_common::api::OutgoingResponse for Response {
+        fn try_into_http_response<T: Default + bytes::BufMut>(
+            self,
+        ) -> Result<http::Response<T>, ruma_common::api::error::IntoHttpError> {
+            match self {
+                Response::Redirect(Redirect { url }) => Ok(http::Response::builder()
+                    .status(http::StatusCode::FOUND)
+                    .header(http::header::LOCATION, url)
+                    .body(T::default())?),
+                Response::Html(HtmlPage { body }) => Ok(http::Response::builder()
+                    .status(http::StatusCode::OK)
+                    .header(http::header::CONTENT_TYPE, "text/html")
+                    .body(ruma_common::serde::slice_to_buf(&body))?),
+            }
+        }
+    }
+
+    #[cfg(feature = "client")]
+    impl ruma_common::api::IncomingResponse for Response {
+        type EndpointError = crate::Error;
+
+        fn try_from_http_response<T: AsRef<[u8]>>(
+            response: http::Response<T>,
+        ) -> Result<Self, ruma_common::api::error::FromHttpResponseError<Self::EndpointError>>
+        {
+            use ruma_common::api::{
+                error::{DeserializationError, FromHttpResponseError},
+                EndpointError,
+            };
+
+            if response.status().as_u16() >= 400 {
+                return Err(FromHttpResponseError::Server(
+                    Self::EndpointError::from_http_response(response),
+                ));
+            }
+
+            if let Some(location) = response.headers().get(http::header::LOCATION) {
```

Comment:
I think we should branch on the status code here, not on the header presence. (and make it an error to if the header is missing with the `FOUND` status code)

Reply:
I think we should check both, the `Location` header only serves a meaning for status code `201` and `3xx` after all.

Reply:
Changed